[REG-1666] Gameplay session docs (#66)
Regression-Games/RegressionDocs: docs/gameplay-sessions/gameplay-sessions-usage.mdx

---
sidebar_label: 'Using Gameplay Sessions'
---

import OverlayPartial from '../partials/_add-overlay-partial.mdx';

# Using Gameplay Sessions

## Preparing the Overlay

First, add the RGOverlayCanvas prefab to your scene, which is the UI that allows us to record and play gameplay
sessions.

> <OverlayPartial />

Author: 🎉

## Recording Gameplay

Recording can be started by clicking the red record button in the overlay under the Regression Games icon.
Once clicked, the SDK will begin recording your gameplay. Click the pulsing red record button again to stop the
recording, at which point all data collected will be saved to your device and uploaded to your Regression Games account.

![Start a recording](img/record.png)

## Playing Back a Recording

A recording can be played back by selecting the menu button underneath the Regression Games icon in the overlay. A file
selector will appear, with a default location of the recordings folder on your device. Navigate into the folder of the
run you'd like to use, select the `data.zip` file, and click "Load Replay".

After the replay is selected, click the Play button underneath the Regression Games icon to begin the playback.

![Open the Gameplay Session Selector](img/overlay_click_load.png)
![Select the Gameplay Session](img/overlay_choose_session.png)

## Advanced Options

> ### Ignoring objects in your Scene

Reviewer: This comment has Not been addressed

> ### Ignoring objects in your Scene

Author: Added!

In certain scenarios, you may want to ignore certain objects in your recordings, such as regions of a level that are
not relevant to the test or behavior of your recording. To ignore these game objects, follow these steps:

1. Click "Add Component"

> 1. Search for and add the `RG Ignore From State` component

Author: Fixed (I believe @abeizer's commit fixes this)

> 1. Search for and add the `RG Ignore From State` component

Reviewer: Yes, I had already marked this point resolved

This will ignore all state from this game object as well as all children of this game object.

### Choosing an FPS for recording

The **Recording Min FPS** option tells the SDK the frequency at which it should record information from the game. For
instance, if the FPS is set to `5`, then 5 times a second the SDK will scrape the game information.

> If the FPS is set `0`, then the SDK will only record information when a

Reviewer: This comment was only partially addressed, we still need a disclaimer that this value should normally stay as 0.

> If the FPS is set `0`, then the SDK will only record information when a

Author: Added, let me know if that clarification is correct?

[**key frame**](/gameplay-sessions/gameplay-sessions-reference#key-frames) is detected.

![Change the FPS at which data is collected from the game](img/base_options_fps.png)

### Changing your recording location

By default, all data is stored to a `unity_videos` directory within your home directory. This can be changed within the
**State Recordings Directory** field.

TO DISCUSS IN PR COMMENTS: How does this work when multiple people are using the project on different machines?

![Change the location where data is saved](img/base_options_location.png)

### Input threshold options

> In some cases, different devices may report spammy key clicks when a key is actually being held down. This threshold

Reviewer 2: I like `key presses`

> In some cases, different devices may report spammy key clicks when a key is actually being held down. This threshold

Author: Fixed by Abby's changes

defines a maximum amount of time that must exist between key clicks to be considered clicks rather than a key being

> held. This defaults to 0.1 seconds, and often should not have to be changed. If you do experience key click spamming,

Reviewer: This comment has Not been addressedThis comment has Not been addressed

> held. This defaults to 0.1 seconds, and often should not have to be changed. If you do experience key click spamming,

Author: Fixed

you can change this value to a higher number within the **Key Held Threshold Seconds** field.

![Change the threshold for held key inputs](img/base_options_threshold.png)

### Game object finder configurations

The SDK traverses the game object tree of a game to collect state relevant for both recording playback and testing.

> In addition to ignore certain objects as described above, there are also a few other ways that this can be configured.

Author: I believe this section has been removed / rearranged so this is no longer relevant

#### Collapsing Renderers

The **Collapse Renderers into Top Level Game Object** checkbox dictates how renderer game objects are collected as part
of the game state. If true, all renderers are collapsed into their top level gameObject. If a gameObject hierarchy
exists that has colliders/renderers/animators/etc at multiple levels, they will all be represented by a single entry in
the state. This defaults to checked as it is normally desired to see each player, car, building, etc as a single entry
in the state. However, it can be useful to set this to false in cases where you want to validate individual render

> bounds, colliders, rigid bodies, and more on individual armatures, weapons, or other components that are children in the

Reviewer: This comment has Not been addressed

> bounds, colliders, rigid bodies, and more on individual armatures, weapons, or other components that are children in the

Author: Fixed!

hierarchy.

#### Include Only On Camera Objects

The **Include Only On Camera Objects** checkbox dictates whether the recorder should collect all objects in the
scene as part of the game state, or only those that appear in the viewport of the camera. This defaults to checked, as
in many cases, objects that are not on camera are not relevant to the test or playback. If you do need to see all

> objects in the scene, uncheck this field.

Reviewer: This comment has Not been addressed

> objects in the scene, uncheck this field.

Author: Added!

#### Collect State from Behaviours

The **Collect State from Behaviours** checkbox dictates whether the recorder should collect fields and property
values from Behaviours attached to the game objects. For recording playback, this is not always needed, but for testing
specific states within your game objects, this is extremely useful. This is unchecked by default, and note that

> enabling this can introduce a performance impact.

Author: Added from Abby's change

![Start a recording](img/record.png)

## Playing Back a Recording

A recording can be replayed by initiating "play" mode from your Unity editor and selecting the menu button below the Regression Games logo.
This will launch a file explorer at the location of the recordings folder on your device.
Navigate into the folder of the recording you'd like to replay, select its `data.zip` file, and then click "Load Replay".
Once you've selected a replay, click the Play button below the Regression Games logo to begin playback.

![Open the Gameplay Session Selector](img/overlay_click_load.png)
![Select the Gameplay Session](img/overlay_choose_session.png)

## Advanced Options

### Choosing an FPS for recording

The `Recording Min FPS` field under `RGOverlayCanvasV2` > `ScreenRecorder` allows you to configure
how often the SDK scrapes the game for state data.
Its default value is 0, which instructs the SDK to only record state data when it detects a
[**key frame**](/gameplay-sessions/gameplay-sessions-reference#key-frames). In most cases, this should be set to 0, as most games will only require state collection on key frames.

> Setting `Recording Min FPS` to a value greater than 0 will cause the SDK to record data at that frequency instead,

Reviewer: ```suggestion
Setting `Recording Min FPS` to a value greater than 0 will cause the SDK to record data at that frequency,
```

> regardless of whether a key frame was detected.

Reviewer: ```suggestion
in addition to recording any key frame that was detected.
```
